test/unit/package_item_test.rb

```diff
@@ -0,0 +1,237 @@
+    assert_nil @item.hs_code
+    assert_nil @item.sku
+    refute_equal @weight, @item.ounces
+    assert_equal @weight, @item.grams
+  end
+
+  def test_initialize_negative_quantity
+    assert_equal 1, PackageItem.new(@name, @weight, @value, -1).quantity
+    assert_equal 1, PackageItem.new(@name, @weight, @value, 0).quantity
+  end
+
+  def test_initialize_weight_mass_object
+    begin
+      @item = PackageItem.new(@name, @mass, @value, @quantity, @options)
+      assert_equal @mass, @item.weight
+      flunk "This code path is broken but passed unexpectedly"
+    rescue NameError
+      skip "This code path is broken"
+    end
+  end
+
+  def test_initialize_weight_default_metric
+    assert_equal @weight, @item.grams
+    refute_equal @weight, @item.ounces
+  end
+
+  def test_initialize_weight_accepts_imperial
+    @item = PackageItem.new(@name, @weight, @value, @quantity, @options.merge(units: :imperial))
+
+    assert_equal @weight, @item.ounces
+    refute_equal @weight, @item.grams
+  end
+
+  def test_initialize_weight_accepts_metric
+    @item = PackageItem.new(@name, @weight, @value, @quantity, @options.merge(units: :metric))
+
+    assert_equal @weight, @item.grams
+    refute_equal @weight, @item.ounces
+  end
+
+  def test_initialize_weight_does_not_accept_strings
+    @item = PackageItem.new(@name, @weight, @value, @quantity, @options.merge(units: "imperial"))
+
+    assert_equal @weight, @item.grams
+    refute_equal @weight, @item.ounces
+  end
+
+  def test_initialize_value_from_cents
+    @item = PackageItem.new(@name, @weight, "1.23", @quantity, @options)
+
+    assert_equal 123, @item.value
+  end
+
+  def test_weight
+    assert_equal @mass, @item.weight
+    assert_instance_of ::Quantified::Mass, @item.weight
+  end
+
+  def test_weight_actual
+    assert_equal @mass, @item.weight(type: :actual)
+    assert_instance_of ::Quantified::Mass, @item.weight(type: :actual)
+  end
+
+  def test_weight_volumetric
+    begin
+      assert_equal :todo, @item.weight(type: :volumetric)
+      assert_instance_of ::Quantified::Mass, @item.weight(type: :volumetric)
+      flunk "This code path is broken but passed unexpectedly"
+    rescue NoMethodError
+      skip "This code path is broken"
+    end
+  end
+
+  def test_weight_dimensional
+    begin
+      assert_equal :todo, @item.weight(type: :dimensional)
+      assert_instance_of ::Quantified::Mass, @item.weight(type: :dimensional)
+      flunk "This code path is broken but passed unexpectedly"
+    rescue NoMethodError
+      skip "This code path is broken"
+    end
+  end
+
+  def test_weight_billable_max_weight_and_volumetric
+    begin
+      assert_equal :todo, @item.weight(type: :billable)
+      assert_instance_of ::Quantified::Mass, @item.weight(type: :billable)
+      flunk "This code path is broken but passed unexpectedly"
+    rescue NoMethodError
+      skip "This code path is broken"
+    end
+  end
+
+  def test_grams_value
+    assert_equal 100, @item.grams
+  end
```

Comment:
I don't think this is part of a bad interface, but simply a mixture of testing paradigms. There's a paradigm where you focus on individual methods and the inputs/outputs of those. Then there's behavioural testing where you have expected behaviours of your classes and test those, which often means testing across methods/accessors/readers/etc. I'm a big fan of the latter, but it's fine to have some of the former too.

What's confusing to me, and the reason for this comment, is that we have `test_initialize_with_all_attributes` test, which tests the state of things after initialization, and that's exactly what this test is also doing. We've taken a behaviour (state immediately after initialization) and split it across the wrong boundary (`attr_reader` / `def foo`). Whether there's logic or not happening in those should be irrelevant to clients of the object. Also, there's nothing stopping us from "upgrading" one of the `attr_reader`s to a manual `def attr; ...stuff...; end` in the future.

Sorry for the wordiness, but hopefully that cleared up my stance :)